sources/metricdog/src/service_check.rs

```diff
@@ -0,0 +1,159 @@
+use crate::error::{self, Result};
+use log::trace;
+use snafu::ResultExt;
+use std::process::Command;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) struct ServiceHealth {
+    /// Whether or not the service is healthy.
+    pub(crate) is_healthy: bool,
+    /// In the event of an unhealthy service, the service's exit code (if found).
+    pub(crate) exit_code: Option<i32>,
+}
+
+pub(crate) trait ServiceCheck {
+    /// Checks the given service to see if it is healthy.
+    fn check(&self, service_name: &str) -> Result<ServiceHealth>;
+}
+
+pub(crate) struct SystemdCheck {}
+
+impl ServiceCheck for SystemdCheck {
+    fn check(&self, service_name: &str) -> Result<ServiceHealth> {
+        if is_ok(service_name)? {
+            return Ok(ServiceHealth {
+                is_healthy: true,
+                exit_code: None,
+            });
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
+        })
+    }
+}
+
+struct Outcome {
+    exit: i32,
+    stdout: String,
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
+        self.exit == 0
+    }
+}
+
+fn systemctl(args: &[&str]) -> Result<Outcome> {
+    trace!("calling systemctl with '{:?}'", args);
+    let output = Command::new("systemctl")
+        .args(args)
+        .output()
+        .context(error::Command {
+            command: "systemctl",
+            args: args.iter().map(|&s| s.to_owned()).collect::<Vec<String>>(),
+        })?;
+    Ok(Outcome {
+        exit: output.status.code().unwrap_or(-1),
+        stdout: String::from_utf8_lossy(output.stdout.as_slice()).into(),
+    })
+}
+
+fn is_active(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-active", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_failed(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-failed", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_ok(service: &str) -> Result<bool> {
+    Ok(!is_failed(service)? && is_active(service)?)
+}
+
+const STATUS_PROPERTY: &str = "ExecMainStatus";
+const PROP_LEN: usize = STATUS_PROPERTY.len();
```

Comment:
It feels like splitting on `=` might be simpler than length math?

Reply:
Depends on whether you find functional or procedural code simpler, I suppose. But here is the split iterator version [b8b5c3a](https://github.com/bottlerocket-os/bottlerocket/compare/153acff977082318fef8fc8e61024b26574c362e..b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187)

Reply:
I don't think split has to be functional like it is in the current revision; I think you're implying that it's harder to read, and I would agree.  We can do something like [pluto](https://github.com/bottlerocket-os/bottlerocket/blob/741750ac742af91014c306183e94dc04285084f7/sources/api/pluto/src/main.rs#L208) or [bottlerocket-release](https://github.com/bottlerocket-os/bottlerocket/blob/741750ac742af91014c306183e94dc04285084f7/sources/bottlerocket-release/src/lib.rs#L75) that's a bit more procedural and clearer - `.next().context(bla)`

Reply:
This seems readable [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737). Slightly different from the reference implementations since I'm checking that the key matches my expected key.